### PR TITLE
KAFKA-6611: PART I, Use JMXTool in SimpleBenchmark

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Use JmxMixin for SimpleBenchmark (will remove the self reporting in #4744), only when loading phase is false (i.e. we are in fact starting the streams app).

2. Reported the full jmx reported metrics in log files, and in the returned data only return the max values: this is because we want to skip the warming up and cooling down periods that will have lower rate numbers, while max represents the actual rate at full speed.

3. Incorporates two other improves to JMXTool: https://github.com/apache/kafka/pull/1241 and https://github.com/apache/kafka/pull/2950

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by guozhangwang** [comment]
@bbejeck @dguy @rodesai for reviews

**Comment by ewencp** [comment]
Just FYI, one of the reasons I wrote the `PushHttpMetricsReporter` (on the java side) and `HttpMetricsCollector` (on the python/system test side) is that while the implementation of the mixin itself isn't necessarily the prettiest, it doesn't have some of the race conditions we had with the `JmxMixin` and it's written to work well as a true mixin for `Service` implementations such that the vast majority of the time, all you need to do to make it work properly is configure the metrics reporter and include `HttpMetricsCollector` in your `Service`'s inheritance hierarchy (e.g. https://github.com/apache/kafka/blob/trunk/tests/kafkatest/services/performance/producer_performance.py#L28 is all that's required for the first service I replaced `JmxMixin` in). It's not a huge amount of code saved, but it's error prone to both remember the right calls and get them invoked in the right place.

There are tradeoffs (e.g. right now `PushHttpMetricsReporter` collects all metrics rather than allowing filtering, may not log as extensively as you'd get from `JMXTool`; on the other hand, already had flexible reporting frequency...), but I also think delivering stats directly and immediately via HTTP rather than collecting logs after the fact (which I think also has some repeat collection issues which the tests might just currently mask/not run into) is going to be a better long term solution and make writing tests simpler.

I'm fine merging this to use the `JmxMixin` version of things given it is already written, but I think we should seriously consider switching away from `JMXTool` in our tests in favor of the http reporter as a better long-term solution.

File: tests/kafkatest/services/monitor/jmx.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def read_jmx_output(self, idx, node):
 
         for name in object_attribute_names:
             aggregates_per_time = []
-            for time_sec in xrange(start_time_sec, end_time_sec + 1):
+            for time_sec in xrange(start_time_sec, end_time_sec + int(self.report_interval/100) + 1):
```
Comment (mjsax) [inline tests/kafkatest/services/monitor/jmx.py]: Not sure why this change? Can you elaborate?

**Comment by bbejeck** [comment]
Overall LGTM.

But given @ewencp comments above, how much of a delay, if any, would we incur removing the self-reporting from the `SimpleBenchmark` and going with `PushHttpMetricsReporter` vs `JmxMixin`?

 I'm not familiar with either the `JmxMixin` or the `PushHttpMetricsReporter` so I don't have a strong opinion on this.

**Comment by guozhangwang** [comment]
Synced offline with @ewencp on `JmxMixin` v.s. `PushHttpMetricsReporter`. And here's my take:

1. In the longer term we would replace the former with the latter for our system tests; however given that may not come in near term, it still makes sense to add this PR (it introduces some tech debt) for now so that in the system test nightly builds we can have some more accurate metrics to detect regressions.

2. For `JMXTool.java` itself, since it is widely used in OS Kafka, we'd be very cautious if we ever wants to deprecate it. Plus for some manual benchmarking tests, it may still makes sense to use the tool for quick-and-dirty metrics collection purposes.

File: tests/kafkatest/services/monitor/jmx.py (hunk 1 above)
Comment (guozhangwang) [reply tests/kafkatest/services/monitor/jmx.py]: The old code assumes that the report interval is always 1 second, and hence hard code it here, whereas I tried to make it configurable and in practice making the reporting interval to be much larger, and hence need to make this change. But we still keep the `+ 1` in case the report interval is configured to be smaller than 1 sec. Does that make sense?

BTW it should be `1000` in the denominator than `100` ..

File: tests/kafkatest/services/monitor/jmx.py [hunk 2]
```diff
@@ -110,7 +110,7 @@ def read_jmx_output(self, idx, node):
                 object_attribute_names = line.strip()[1:-1].split("\",\"")[1:]
                 continue
             stats = [float(field) for field in line.split(',')]
-            time_sec = int(stats[0]/1000)
+            time_sec = int(stats[0]/self.report_interval)
```
Comment (mjsax) [inline tests/kafkatest/services/monitor/jmx.py]: If the reporting interval itself is increase, `stats[0]` would just have a different value, but this conversion from `ms` to `sec` should still be `/1000` if I understand the code correctly?

File: tests/kafkatest/services/monitor/jmx.py (hunk 1 above)
Comment (mjsax) [reply tests/kafkatest/services/monitor/jmx.py]: Not sure if we need to change the end of the range, but rather the "step size" of the `xrange` to change from `1` to `report_interval` ?

File: tests/kafkatest/services/kafka/kafka.py [hunk 3]
```diff
@@ -79,7 +79,7 @@ def __init__(self, context, num_nodes, zk, security_protocol=SecurityConfig.PLAI
         :type topics: dict
```
Comment (rodesai) [inline tests/kafkatest/services/kafka/kafka.py:78]: Not really related to this pr - I was looking at the surrounding code and noticed that the `server_prop_overides arg` is defaulted to `[]`. Its probably fine since its not being written, but you almost never want that since every instance will share the same global list. Might want to fix that while you're here.

**Review by rodesai — APPROVED** [review]
The changes look good to me. One comment unrelated to this PR inline.

File: streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java [hunk 4]
```diff
@@ -678,9 +679,11 @@ public void apply(Integer key, V value) {
             } else {
                 System.err.println("Unknown value type in CountDownAction");
             }
+            /*
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java]: Should those line get removed?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/perf/SimpleBenchmark.java]: ack.

File: tests/kafkatest/services/kafka/kafka.py (hunk 3 above)
Comment (guozhangwang) [reply tests/kafkatest/services/kafka/kafka.py:78]: @rodesai what do you mean by the "global list"?

Comment (rodesai) [reply tests/kafkatest/services/kafka/kafka.py:78]: When the interpreter processes the function def it will initialize one list that it passes to every instance of KafkaService thats constructed without specifying its own server_prop_overrides. Here's an illustrative example:
```code
>>> class A(object):
...     def __init__(self, a, b=[]):
...         b.append(a)
...         self.b = b
...
>>> print A(1).b
[1]
>>> print A(2).b
[1, 2]
>>> print A(3).b
[1, 2, 3]
```

Comment (guozhangwang) [reply tests/kafkatest/services/kafka/kafka.py:78]: Got it, thanks! Please let me know if the fix makes sense to you.

Comment (rodesai) [reply tests/kafkatest/services/kafka/kafka.py:78]: This still has the same problem, though its more explicit. Not sure if that was the intent. The standard approach is something like this
```{code}
def __init__(self, ..., server_prop_overrides=None, ...):
    if server_prop_overrides is None:
        server_prop_overrides = []
```

Comment (guozhangwang) [reply tests/kafkatest/services/kafka/kafka.py:78]: Ah got it. I will try this in the code:

```
if server_prop_overides is None:
            self.server_prop_overides = []
else:
            self.server_prop_overides = server_prop_overides
```

**Comment by guozhangwang** [comment]
Succeeded streams benchmark test: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1548/

File: tests/kafkatest/benchmarks/streams/streams_simple_benchmark_test.py [hunk 5]
```diff
@@ -35,7 +35,7 @@ def __init__(self, test_context):
         self.num_threads = 1
 
     @cluster(num_nodes=9)
-    @matrix(test=["produce", "consume", "count", "processstream", "processstreamwithsink", "processstreamwithstatestore", "processstreamwithcachedstatestore", "kstreamktablejoin", "kstreamkstreamjoin", "ktablektablejoin", "yahoo"], scale=[1, 3])
+    @matrix(test=["count", "processstream", "processstreamwithsink", "processstreamwithstatestore", "processstreamwithcachedstatestore", "kstreamktablejoin", "kstreamkstreamjoin", "ktablektablejoin"], scale=[1, 3])
```
Comment (guozhangwang) [inline tests/kafkatest/benchmarks/streams/streams_simple_benchmark_test.py:38]: This is intentional: I removed produce / consume and yahoo since they do not have streams app and hence no jmx metrics to collect, I'm adding some more logic to re-enable them in https://github.com/apache/kafka/pull/4744

**Comment by guozhangwang** [comment]
@mjsax @bbejeck @vvcephei @dguy call for reviews.

Please note that the changes in `JMXTool.java` was cherry-picked from some existing PRs from https://issues.apache.org/jira/browse/KAFKA-4620?jql=project%20%3D%20KAFKA%20AND%20resolution%20%3D%20Unresolved%20AND%20text%20~%20%22JMXTool%22%20ORDER%20BY%20priority%20DESC

**Comment by vvcephei** [comment]
@guozhangwang This looks good to me.

File: tests/kafkatest/benchmarks/streams/streams_simple_benchmark_test.py [hunk 6]
```diff
@@ -93,11 +95,18 @@ def test_simple_benchmark(self, test, scale):
...
         for num in range(0, scale):
             for key in data[num]:
                 final[key + str(num)] = data[num][key]
-        
+
+            for key in sorted(self.driver[num].jmx_stats[0]):
+                self.logger.info("%s: %s" % (key, self.driver[num].jmx_stats[0][key]))
+
+            final["jmx-avg" + str(num)] = self.driver[num].average_jmx_value
+            final["jmx-max" + str(num)] = self.driver[num].maximum_jmx_value
```
Comment (mjsax) [inline tests/kafkatest/benchmarks/streams/streams_simple_benchmark_test.py:110]: Just for my own education: average and max of what?

File: tests/kafkatest/services/monitor/jmx.py [hunk 7]
```diff
@@ -70,14 +70,14 @@ def check_jmx_port_listening():
...
-                          self.jmx_class_name())
+        cmd = "%s %s " % (self.path.script("kafka-run-class.sh", use_jmxtool_version), self.jmx_class_name())
         cmd += "--reporting-interval 1000 --jmx-url service:jmx:rmi:///jndi/rmi://127.0.0.1:%d/jmxrmi" % self.jmx_port
         cmd += " --wait"
         for jmx_object_name in self.jmx_object_names:
             cmd += " --object-name %s" % jmx_object_name
+        cmd += " --attributes "
         for jmx_attribute in self.jmx_attributes:
-            cmd += " --attributes %s" % jmx_attribute
+            cmd += "%s," % jmx_attribute
```
Comment (mjsax) [inline tests/kafkatest/services/monitor/jmx.py:80]: The additional comma after the last attribute is no an issue?

File: tests/kafkatest/services/performance/streams_performance.py [hunk 8]
```diff
@@ -31,6 +31,54 @@ def __init__(self, test_context, kafka, numrecs, load_phase, test_name, num_thre
                                                             test_name,
                                                             num_threads)
 
+        self.load_phase = load_phase
+
+        if self.load_phase == "false":
+            JmxMixin.__init__(self, num_nodes=1,
```
Comment (mjsax) [inline tests/kafkatest/services/performance/streams_performance.py]: nit: should `num_nodes=1` be it's own line?

**Review by mjsax — COMMENTED** [review]
Wondering why this PR has 60 commits... Overall LGTM

**Comment by mjsax** [comment]
Can you link to the PR for KAFKA-4620 ? The ticket does not contain a link either....

File: tests/kafkatest/benchmarks/streams/streams_simple_benchmark_test.py (hunk 6 above)
Comment (guozhangwang) [reply tests/kafkatest/benchmarks/streams/streams_simple_benchmark_test.py:110]: This is to calculate the avg and max of all the collected metrics. In this PR I'm collecting poll / process / commit rate and latency, and compute the avg / max of all of them

In the follow-up PR (https://github.com/apache/kafka/pull/4744) I'm collecting consumer metrics for comparison.

Sorry I have do do this series in not-so-independent PRs, as I do not want to add a too large one, but separating them into steps is also a bit tricky.

I have also updated the description of this PR. Hope that helps understanding this change better.

File: tests/kafkatest/services/monitor/jmx.py (hunk 7 above)
Comment (guozhangwang) [reply tests/kafkatest/services/monitor/jmx.py:80]: Yes, as in the JMXTool java code I intentionally handles that part: https://github.com/apache/kafka/pull/4650/files#diff-096d07271294637c2e52b68dd1038c44R104

File: tests/kafkatest/services/performance/streams_performance.py (hunk 8 above)
Comment (guozhangwang) [reply tests/kafkatest/services/performance/streams_performance.py]: ack.

**Comment by guozhangwang** [comment]
> Can you link to the PR for KAFKA-4620 ? The ticket does not contain a link either....

Sorry for the mis-communication, this PR does not actually fix KAFKA-4620. It fixed 1972 and 3591 though, the long commit history is also because I cherry-picked their commits into this trace.

**Review by mjsax — APPROVED** [review]
LGTM.
